source/utils/adios_remote_read/adios_remote_read.cpp

```diff
@@ -0,0 +1,721 @@
+ */
+#include <chrono>
+#include <fstream>
+#include <getopt.h>
+#include <iostream>
+#include <stdio.h>
+#include <string.h>
+#include <vector>
+
+#include "adios2.h"
+
+#if ADIOS2_USE_MPI
+#include <algorithm>
+#include <mpi.h>
+#endif
+bool DEBUG = false;
+enum test_cases
+{
+    DIM1,
+    DIM3X,
+    DIM3Y,
+    DIM3Z,
+    DIM3PLANEYZ,
+    DIM3PLANEXY,
+    DIM3PLANEXZ
+};
+void read1D(int nproc, int rank, const std::string &filename, const int nsteps, adios2::IO &io,
+            std::vector<std::string> &variables, int output_length);
+/* test 2 and 3
+ * read 1 or many 1D variables
+ */
+void read3D(int nproc, int rank, const std::string &filename, const int nsteps, adios2::IO &io,
+            std::vector<std::string> &variables, int direction, double ratio,
+            int output_line_length);
+/* test 5
+ * A 3D subset from 3D variable
+ */
+
+void read3DPlane(int nproc, int rank, const std::string &filename, const int nsteps, adios2::IO &io,
+                 std::vector<std::string> &variables, int direction, double ratio,
+                 int output_line_length);
+
+std::string &getOutputString(int rank, int output_line_length, std::string &out,
+                             std::vector<double> &data);
+
+/* test 5
+ * A 3D subset from 3D variable
+ */
+
+void read1D(int nproc, int rank, const std::string &filename, const int nsteps, adios2::IO &io,
+            std::vector<std::string> &required_variables, double ratio, int output_line_length)
+{
+    unsigned int startX;
+    unsigned int countX;
+
+    try
+    {
+        std::chrono::time_point<std::chrono::system_clock> start;
+        std::chrono::time_point<std::chrono::system_clock> end;
+        start = std::chrono::system_clock::now();
+
+        adios2::Engine reader = io.Open(filename, adios2::Mode::Read);
+
+        /* get variables with 1D shape */
+
+        std::string out;
+        for (size_t step = 0; step < nsteps; step++)
+        {
+            reader.BeginStep();
+            auto available_variables = io.AvailableVariables(true);
+            std::vector<std::string> available_variables_str;
+            for (const auto &v : available_variables)
+                available_variables_str.push_back(v.first);
+            std::vector<std::string> vars_intersection;
+            if (!required_variables.empty())
+            {
+                std::sort(available_variables_str.begin(), available_variables_str.end());
+
+                std::set_intersection(available_variables_str.begin(),
+                                      available_variables_str.end(), required_variables.begin(),
+                                      required_variables.end(),
+                                      std::back_inserter(vars_intersection));
+            }
+            else
+            {
+                vars_intersection = available_variables_str;
+            }
+
+            for (auto const &var_name : vars_intersection)
```

Comment:
Why not just parse over the `required_variables`?

Reply:
In the case that --variables parameter is not specified, It think the program should try to read all available variables.
Additionally, Input parameters not always correct.

Reply:
If input parameters are incorrect, the inquire variable function will return a null variable and you check for this so you would skip this variable anyway. So I would ask for available variables only if the provided variables is empty (I would also add a comment to be clear how the function should behave for any valid input). There is no point to do the intersection, it just makes the code harder to read.